### PR TITLE
"noqa": Add support for ignoring template (TMP) and parse (PRS) errors

### LINKED ISSUE #1959: noqa is ignored for jinja templated lines
## Expected Behaviour
Line with `noqa: TMP` should be ignored (despite of evaluation error)

## Observed Behaviour
trying to lint airflow sql-template for AWS Athena query
setting up inline `-- noqa` or `--noqa: TMP` for jinja templated line not silenting templating error (typecasting error due to unable to pass datetime object while linting into template context):
```
== [transform/airflow/dags/queries/sfmc/player_balance.sql] FAIL
L:   0 | P:   0 |  TMP | Unrecoverable failure in Jinja templating: unsupported operand type(s) for -: 'int' and 'datetime.timedelta'. Have you configured your variables?
                       | https://docs.sqlfluff.com/en/latest/configuration.html
```

## Steps to Reproduce
templated file:
```sql
select *, row_number() over (partition by player_id order by balance_change_date desc)  as rnk
from raw
where
    balance_change_date >= cast(from_iso8601_timestamp('{{ execution_date - macros.timedelta(hours=2, minutes=10) }}') as timestamp)  and  --noqa: TMP
    balance_change_date < cast(from_iso8601_timestamp('{{ next_execution_date - macros.timedelta(minutes=10) }}') as timestamp) --noqa: TMP
```
run:
```bash
sqlfluff lint transform/airflow/dags/queries/sfmc/player_balance.sql
```

## Dialect
postgres (used for AWS Athena)

## Version
datalake % sqlfluff --version
sqlfluff, version 0.8.1
datalake % python3 --version
Python 3.9.8

## Configuration
```ini
# tox.ini
[sqlfluff]
templater = jinja
output_line_length = 180
exclude_rules = L011,L012,L022,L031,L034
dialect = postgres

[sqlfluff:rules]
max_line_length = 120

[sqlfluff:templater:jinja]
library_path = operation/deploy/lint
apply_dbt_builtins = false

[sqlfluff:templater:jinja:context]
ds = 2021-11-11
ds_nodash = 20211111
start_date = 2021-11-11
end_date = 2021-11-11
interval = 1
# passed as int due to inabliity to pass datetime obkject 
data_interval_start = 1636588800
data_interval_end = 1636588800
```

```python
# operation/deploy/lint/macro.py
from datetime import datetime, timedelta  # noqa: F401

import dateutil  # noqa: F401
```

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #1959, #2118

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/core/linter/common.py [hunk 1]
```diff
@@ -65,6 +66,7 @@ class ParsedString(NamedTuple):
     templated_file: TemplatedFile
     config: FluffConfig
     fname: str
+    source_str: str
```
Comment (barrywhart) [inline src/sqlfluff/core/linter/common.py:69]: Added `source_str` fields to these two structures so the linter has access to the raw source in case of template and/or parse failures. This allows the linter to find `noqa` comments even when these failures occur.

File: requirements_dev.txt [hunk 2]
```diff
@@ -1,5 +1,7 @@
 # Install with -U to keep all requirements up-to-date
 # code linting and formatting
+# Black 22.1 has a fatal bug, so use the latest previous for now
+black==21.12b0
```
Comment (barrywhart) [inline requirements_dev.txt:4]: I first tried `black!=22.1`, which I think **would've** worked, except Black has **never** had a "stable" release until now. They are all beta, so if we exclude the one and only stable release, `pip` refuses to install any of the **unstable** releases. Ironically, the first ever "stable" release is fatally broken. 🙄🤣

Prior to this change, 22.1 was giving the same error on every file it processed:
```
linting run-test: commands[0] | flake8
./setup.py:0:1: BLK999 Unexpected exception: unsupported operand type(s) for /: 'tuple' and 'str'
./util.py:0:1: BLK999 Unexpected exception: unsupported operand type(s) for /: 'tuple' and 'str'
./plugins/sqlfluff-templater-dbt/setup.py:0:1: BLK999 Unexpected exception: unsupported operand type(s) for /: 'tuple' and 'str'
./plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:0:1: BLK999 Unexpected exception: unsupported operand type(s) for /: 'tuple' and 'str'
./plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/__init__.py:0:1: BLK999 Unexpected exception: unsupported operand type(s) for /: 'tuple' and 'str'
./plugins/sqlfluff-templater-dbt/test/templater_test.py:0:1: BLK999 Unexpected exception: unsupported operand type(s) for /: 'tuple' and 'str'
./plugins/sqlfluff-templater-dbt/test/rules_test.py:0:1: BLK999 Unexpected exception: unsupported operand type(s) for /: 'tuple' and 'str'
```

File: src/sqlfluff/core/linter/linter.py [hunk 3]
```diff
@@ -556,6 +585,21 @@ def lint_parsed(
...
+                # requires access to the parse tree, and because of the failure,
+                # we don't have a parse tree).
+                ignore_buff, ignore_violations = cls.extract_ignore_mask_source(
+                    parsed.source_str,
+                    [
+                        lm
+                        for lm in parsed.config.get("dialect_obj").lexer_matchers
+                        if lm.name == "inline_comment"
+                    ][0],
+                    [r.code for r in rule_set],
```
Comment (tunetheweb) [inline src/sqlfluff/core/linter/linter.py:600]: At no point do we look for "noqa" here. Should we be?

File: src/sqlfluff/core/linter/linter.py [hunk 4]
```diff
@@ -556,6 +585,21 @@ def lint_parsed(
...
+                )
+                violations += ignore_violations
```
Comment (tunetheweb) [inline src/sqlfluff/core/linter/linter.py:602]: Why are you adding the violations on? Shouldn't you be removing them because they are ignorable violations?

File: test/core/linter_test.py [hunk 5]
```diff
@@ -708,18 +708,42 @@ def test_linter_noqa_with_templating():
...
+def test_linter_noqa_template_errors():
+    """Similar to test_linter_noqa, but uses templating (Jinja)."""
+    lntr = Linter(
+        config=FluffConfig(
+            overrides={
+                "templater": "jinja",
+            }
+        )
+    )
+    sql = """select * --noqa: TMP
```
Comment (tunetheweb) [inline test/core/linter_test.py:720]: why do we need this "noqa" comment here?

File: src/sqlfluff/core/linter/linter.py (hunk 3 above)
Comment (barrywhart) [reply src/sqlfluff/core/linter/linter.py:600]: We don't need to. The new `extract_ignore_mask_source()` (called at line 593 above) looks for the `noqa` comments.

File: src/sqlfluff/core/linter/linter.py (hunk 4 above)
Comment (barrywhart) [reply src/sqlfluff/core/linter/linter.py:602]: This line handles the case where the `noqa` comment is incorrectly formatted. Basically, if the "noqa" comments don't "parse" correctly.

The important thing here is that we're assigning to `ignore_buffer`. That is the list of parsed "noqa" comments. It'll be applied later. See line 569 in this same file -- that's the normal workflow for assigning to `ignore_buffer`. There are two cases:
* Normal case: Successful template and parse (line 569)
* Error case: Templating and/or parsing fails (here)

In each case, we assign to `ignore_buff`, which is the list of "noqa`" directives. They are actually applied later. Here, all we need to do is build that list and pass it on.

File: test/core/linter_test.py (hunk 5 above)
Comment (barrywhart) [reply test/core/linter_test.py:720]: When templating fails due to an undefined variable, Jinja throws an exception but doesn't tell us what line it happened on. In this PR, I updated it so a templating error is reported with a line number of 1. Since we don't know where the error occurred, this is better than nothing. A user will see the error reported at line 1, then they'll know they need to add a "noqa" comment at line 1. This whole workflow feels a little weird -- as you pointed out in one of the original issues, they could (and probably should) just fix their templating variables. But if somebody wants to ignore these errors, now they can.

File: requirements_dev.txt [hunk 6]
```diff
@@ -2,7 +2,8 @@
 # code linting and formatting
 flake8
 flake8-docstrings
-flake8-black
+# Black 22.1 has a fatal bug when used with flake8-black, so skip it for now.
+#flake8-black
```
Comment (tunetheweb) [inline requirements_dev.txt:6]: This works better (it's about the one combination you DIDN'T try!):

```suggestion
# Black 22.1 has a fatal bug when used with flake8-black, so freeze it for now.
black<=21.12b0
flake8-black
```

I've also pulled this into a separate PR (#2513) that we should merge first.

File: test/core/linter_test.py (hunk 5 above)
Comment (tunetheweb) [reply test/core/linter_test.py:720]: OK so you need it on BOTH line 1 AND the actual failure line by the looks of things.

This looks like something we should mention in https://docs.sqlfluff.com/en/stable/rules.html#inline-ignoring-errors and

**Comment by tunetheweb** [comment]
Change looks good once you address those last two comments.

I also feel we should merge #2513 first to differentiate the black fix from this PR.

File: test/core/linter_test.py (hunk 5 above)
Comment (barrywhart) [reply test/core/linter_test.py:720]: In some cases, yes. It depends on the error. In this case, it was something being "undefined". There are other kinds of errors (e.g. the template code raises an exception or has a syntax error, e.g. missing parenthesis. I'm hesitant to mention it in the docs, because it's situation dependent. If someone needs to do it, they'll know they need to do it because they'll run `sqlfluff lint` and see the message refers to line 1. This seems self explanatory to me. Have you (can you) try linting the above example using this branch, but first remove all the `--noqa` comments? I think you'll see what I mean. If you still feel like the doc change is needed, I'm happy to add it then.

Comment (tunetheweb) [reply test/core/linter_test.py:720]: Yeah that's what I ran to get my mind sorted.

OK if it's not consistent then happy to leave for now. I agree `lint` does point you in that direction anyway.

Resolving.

**Comment by barrywhart** [comment]
@tunetheweb: BTW, I removed the Black-related changes from this PR. Hope we get that figured out soon, or they release an upstream fix. What a mess...

**Comment by tunetheweb** [comment]
> @tunetheweb: BTW, I removed the Black-related changes from this PR. Hope we get that figured out soon, or they release an upstream fix. What a mess...

It's actually not black, it's flake8-black and they've released a fix.

But looks like our dbt tests hang on new black anyway. Not sure what that's about but might leave it for now and try again later next month once things settle down.

**Comment by barrywhart** [comment]
Slightly related, some of the challenges with Black make me wonder if we could drop support for older dbt versions (i.e. prior to dbt 1.0.0). IIRC, Alan said there's very little usage now, and supporting multiple versions of a large package like dbt makes **our** version considerations considerably more complex. I'll bring this up at the Tuesday meeting.

**Comment by tunetheweb** [comment]
> Slightly related, some of the challenges with Black make me wonder if we could drop support for older dbt versions (i.e. prior to dbt 1.0.0). IIRC, Alan said there's very little usage now, and supporting multiple versions of a large package like dbt makes **our** version considerations considerably more complex. I'll bring this up at the Tuesday meeting.

Maybe open an issue for this too so others can comment too? Would be interesting to see feedback before Tuesday if we get any.

**Review by tunetheweb — APPROVED** [review]
